k2/python/k2/dense_fsa.py

```diff
@@ -0,0 +1,53 @@
+# Copyright (c)  2020  Mobvoi Inc.        (authors: Fangjun Kuang)
+
+import _k2
+import numpy as np
+import torch
+
+
+def dense_fsa(log_probs: torch.Tensor,
+              input_lengths: torch.Tensor) -> _k2.DenseFsaVec:
+    '''Construct a DenseFsa from neural net log-softmax outputs.
+
+    Args:
+      log_probs:
+        A 3-D tensor of dtype ``torch.float32`` with shape ``(N, T, C)``, where
+        ``N`` is the batch size, ``T`` the maximum input length, and ``C`` the
+        number of output classes.
```

Comment:
This is not quite right, the format is more general.  Instead of `input_lengths`, we'll have a list of (sequence_index,start_time, duration).
And you haven't made much of an effort here to avoid loops!  Loops that access the GPU, in particular, are not
a great idea.  If you need to do small writes, better to do them on the CPU and transfer things to the GPU in one operation.  I'm open to adding some code at the C++ level to do this formatting, also.  (But that's not urgent, we can leave it for later, assuming we can't find a way to do it efficiently in Python).

Reply:
I was following: https://github.com/k2-fsa/k2/issues/175#issuecomment-702571462
> Note, something would need to be done in Python in general, when constructing it, to add the - infinitys in the right place. (This needs to be done at the Python level, using the base toolkit, so that autograd will work).

I don't think `autograd` plays a role here. We'll calculate the gradient via arc maps manually.

I agree it is more efficient to do this in C++.  I'll move the formatting process to C++.

Reply:
>  Instead of input_lengths, we'll have a list of (sequence_index,start_time, duration).

Do you mean we should also pass the supervision as an argument? Without supervisions, I cannot understand
how to make use of `start_time, duration`. Could you give some hints?

Reply:
Perhaps you were following the "regular" case there.  The regular case was more regular than you realized, I think: i just meant completely rectangular, no raggedness.
The irregular case may have zero or more than one supervision segments per sequence.

Regarding autograd: it still plays a role.  The arc_maps will propagate the gradient back to the special matrix (containing -infinitys and zeros as well as scores) that we construct, but (a) actually in most instances this will happen automatically using PyTorch autograd, because we always create a "parallel" version of the scores, separate from the FSA, and that's used to handle the gradients; and (b) the autograd back from the special matrix containing -infinity's, to the original matrix of scores, will rely on the PyTorch autograd.  [Incidentally, the arc_map will rely on a linearized/flattened version of the special matrix, and we will assume that matrix is contiguous.]
We should ensure that the copying of the scores from the original nnet output into the special matrix is done as a single indexing operation: this will be fastest in both forward and backward passes.

Reply:
RE autograd: I reread the pseudocode from https://github.com/k2-fsa/k2/issues/175. Yes, you're right.
I'll try to figure out how to use PyTorch's index method to do the formatting process.

Reply:
> Loops that access the GPU, in particular, are not
a great idea. 

Fixed. Please have a review.